src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java

```diff
@@ -1186,9 +1263,26 @@ public void doLogout(StaplerRequest req, StaplerResponse rsp) throws IOException
         OicSession oicSession = OicSession.getCurrent();
         if (oicSession != null) {
             // session will be invalidated but we still need this data for our redirect.
-            req.setAttribute(ID_TOKEN_REQUEST_ATTRIBUTE, oicSession.getIdToken());
             req.setAttribute(STATE_REQUEST_ATTRIBUTE, oicSession.getState());
         }
+
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = User.get2(authentication);
+
+        Assert.notNull(user, "User must not be null");
+
+        OicCredentials credentials = user.getProperty(OicCredentials.class);
+
+        if (credentials != null) {
+            if (this.logoutFromOpenidProvider && !Strings.isNullOrEmpty(this.endSessionEndpoint)) {
+                // This ensures that token will be expired at the right time with API Key calls, but no refresh can be
+                // made.
+                user.addProperty(new OicCredentials(null, null, null, credentials.getExpiresAtMillis()));
```

Comment:
Invalidating the refresh token here seems to cause jenkins to fall back to `lastGrantedAuthorities` for the user when doing token based access.  This means (afaict in my testing) that tokens will still be effectively valid even if  that access was removed at the idp level.  The new authorities are only propagated to jenkins if the user (whose access was removed) attempts to log back in again.

If the `refreshToken` is left persisted in the user properties then it will be used to refresh the API token (once the id token as expired) and at that point it will get the updated authorities for the user associated with the token.  For example, if their access to the particular jenkins instance was updated at the IdP then this will correctly get reflected in their jenkins user's `grantedAuthorities`

Reply:
hmm, it seems that if the refresh token is saved then the next time the user tried to log directly into the instance then that token will be used and the credentials refreshed and the user will not be prompted to log back in at the idp, which doesn't really seem ideal either.

Reply:
> Invalidating the refresh token here seems to cause jenkins to fall back to `lastGrantedAuthorities` for the user when doing token based access. This means (afaict in my testing) that tokens will still be effectively valid even if that access was removed at the idp level. The new authorities are only propagated to jenkins if the user (whose access was removed) attempts to log back in again.
> 
> If the `refreshToken` is left persisted in the user properties then it will be used to refresh the API token (once the id token as expired) and at that point it will get the updated authorities for the user associated with the token. For example, if their access to the particular jenkins instance was updated at the IdP then this will correctly get reflected in their jenkins user's `grantedAuthorities`

This is only true if "strict expiration check" is disabled - which is default at the moment. Once it's changed to require refresh token by default, user will be greeted with 403.

Reply:
> hmm, it seems that if the refresh token is saved then the next time the user tried to log directly into the instance then that token will be used and the credentials refreshed and the user will not be prompted to log back in at the idp, which doesn't really seem ideal either.

I'm not sure why do you see this a problem. This is exactly the purpose of refresh token. However, if the user logs out at IdP, the refresh token will be invalidated and the user will be greeted with 403/redirect to login.